### PR TITLE
tcp: fail instantly if local port is unbound

### PR DESCRIPTION
**Pull request by bzoz**
On Windows when connecting to an unavailable port, the connect() will retry for 2s, even on loopback devices. This uses a call to WSAIoctl to make the connect() call fail instantly on local connections.

### CODE REVIEW

**Comment by bzoz** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1939/

**Comment by bzoz** [comment]
libuv-in-node: https://ci.nodejs.org/view/libuv/job/libuv-in-node/138/

File: src/win/tcp.c [hunk 1]
```diff
@@ -792,6 +810,24 @@ static int uv_tcp_try_connect(uv_connect_t* req,
...
+    WSAIoctl(handle->socket,
+             SIO_TCP_INITIAL_RTO,
+             &retransmit_ioctl,
+             sizeof(retransmit_ioctl),
+             NULL,
+             0,
+             &bytes,
+             NULL,
+             NULL);
+  }
```
[the remark this answers is not in the document]
Comment (bzoz) [reply src/win/tcp.c:829]: I think it is not critical. If this fails for whatever reason the only result is that connecting to nonexistent localhost port will take ~2s.

**Comment by bzoz** [comment]
Updated & rebased, PTAL. CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1943/

File: src/win/tcp.c [hunk 2]
```diff
@@ -750,13 +750,31 @@ int uv_tcp_read_start(uv_tcp_t* handle, uv_alloc_cb alloc_cb,
...
+static int uv__is_loopback(const struct sockaddr_storage* storage) {
+  const struct sockaddr_in* in4;
+  const struct sockaddr_in6* in6;
+  if (storage->ss_family == AF_INET) {
+    in4 = (const struct sockaddr_in*) storage;
+    return in4->sin_addr.S_un.S_un_b.s_b1 == 127;
+  }
+  if (storage->ss_family == AF_INET6) {
+    in6 = (const struct sockaddr_in6*) storage;
+    for (int i=0; i<7; ++i) {
```
Comment (bnoordhuis) [inline src/win/tcp.c:762]:
```suggestion
  const struct sockaddr_in* in4;
  const struct sockaddr_in6* in6;
  int i;

  if (storage->ss_family == AF_INET) {
    in4 = (const struct sockaddr_in*) storage;
    return in4->sin_addr.S_un.S_un_b.s_b1 == 127;
  }
  if (storage->ss_family == AF_INET6) {
    in6 = (const struct sockaddr_in6*) storage;
    for (i = 0; i < 7; ++i) {
```

**Comment by bzoz** [comment]
Looks like on OS-es older than Win10/Server 2016 the change makes the connect take about 2min before timing out. Updated, now the change only affects supported systems.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1948/

**Comment by bzoz** [comment]
Looks like this only works on Windows 10.0.16299 and later: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1954/

**Comment by bzoz** [comment]
Updated, PTAL

**Comment by bzoz** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/2005/

**Comment by bzoz** [comment]
landed in https://github.com/libuv/libuv/commit/99e88edf736827a26e8ce9d87da972f41ce828d9

**Pull request closed without merging**